File: dued-rs/src/display.rs
use std::path::Path;

use serde_json::Value;

pub fn print_value(data: &Value) {
    if let Some(obj) = data.as_object() {
        if obj.contains_key("report") && obj.contains_key("files") {
            print_analyze(data);
            return;
        }
        if obj.contains_key("symbols") && obj.contains_key("files") && obj.contains_key("hollow") && !obj.contains_key("edges") {
            print_dead(data);
            return;
        }
        if obj.contains_key("blast_radius") || obj.contains_key("effects") && obj.contains_key("symbols") {
            print_slice(data);
            return;
        }
        if obj.contains_key("clones") && obj.contains_key("clusters") {
            print_cluster(data);
            return;
        }
        if obj.contains_key("summary") && obj.contains_key("coupling") || obj.contains_key("bus_factor") {
            print_history(data);
            return;
        }
        if obj.contains_key("reading_order") {
            print_brief(data, None);
            return;
        }
    }
    if let Some(arr) = data.as_array() {
        print_list(arr);
        return;
    }
    println!("{}", serde_json::to_string_pretty(data).unwrap());
}

pub fn print_analyze(data: &Value) {
    println!("Scan complete");
    println!();
    kv("files", &data["files"]);
    kv("symbols", &data["symbols"]);
    kv("edges", &data["edges"]);
    kv("parsed this run", &data["parsed"]);
    kv("reused files", &data["reused"]);
    kv("issues", &data["issues"]);
    kv("hollow stubs", &data["hollow"]);
    kv("clones", &data["clones"]);
    kv("mismatches", &data["mismatches"]);
    kv("model", &data["model"]);
    kv("elapsed seconds", &data["elapsed_seconds"]);
    if let Some(inv) = data.get("inventory") {
        if let Some(langs) = inv.get("languages").and_then(|v| v.as_array()) {
            let text = langs
                .iter()
                .filter_map(|l| Some(format!("{} ({} files)", l["language"].as_str()?, l["n"])))
                .collect::<Vec<_>>()
                .join(", ");
            if !text.is_empty() {
                println!("  {:<18} {text}", "languages");
            }
        }
    }
    if data["git"]["enabled"].as_bool() == Some(true) {
        println!("  {:<18} {} commits", "git", data["git"]["commits"]);
    }
    println!();
    if let Some(report) = data["report"].as_str() {
        let html = Path::new(report).join("report.html");
        println!("Reports");
        println!("  HTML     {}", html.display());
        println!("  folder   {report}");
        println!();
        println!("Explore");
        println!("  open the HTML file in a browser");
        println!("  dued report              re-print this summary from disk");
        println!("  dued rank                reading order");
        println!("  dued issues              flagged problems");
        println!("  dued dead                unused symbols and files");
        println!("  dued names               name-health flags");
        println!("  dued cluster             clones");
        println!("  dued slice <symbol>      behavior slice");
    }
}

pub fn print_brief(data: &Value, html: Option<&Path>) {
    println!("dued report");
    println!();
    kv("repo", &data["repo"]);
    kv("files", &data["files"]);
    kv("symbols", &data["symbols"]);
    if let Some(langs) = data["languages"].as_array() {
        let text = langs
            .iter()
            .filter_map(|l| Some(format!("{}={}", l["language"].as_str()?, l["n"])))
            .collect::<Vec<_>>()
            .join(", ");
        println!("  {:<18} {text}", "languages");
    }
    println!();
    println!("Reading order");
    if let Some(order) = data["reading_order"].as_array() {
        for (i, item) in order.iter().enumerate() {
            println!(
                "  {:>2}. {}::{}  —  {}",
                i + 1,
                item["relpath"].as_str().unwrap_or(""),
                item["name"].as_str().unwrap_or(""),
                item["why"].as_str().unwrap_or("")
            );
        }
    }
    println!();
    if let Some(issues) = data["issues"].as_array() {
        println!("Issues ({})", issues.len());
        for item in issues.iter().take(15) {
            let loc = match (item["relpath"].as_str(), item["name"].as_str()) {
                (Some(p), Some(n)) if !n.is_empty() => format!("{p}::{n}"),
                (Some(p), _) => p.to_string(),
                _ => String::new(),
            };
            println!(
                "  {:<18} {}  {}",
                item["kind"].as_str().unwrap_or(""),
                loc,
                item["detail"].as_str().unwrap_or("")
            );
        }
        println!();
    }
    if let Some(html) = html {
        println!("HTML report");
        println!("  {}", html.display());
        println!("  open that file in a browser to search and sort the full index");
        println!("  JSON tables are in the data/ folder next to the HTML file");
    }
}

pub fn print_dead(data: &Value) {
    let symbols = data["symbols"].as_array().cloned().unwrap_or_default();
    let files = data["files"].as_array().cloned().unwrap_or_default();
    let hollow = data["hollow"].as_array().cloned().unwrap_or_default();
    println!("Dead code");
    println!("  unused symbols  {}", symbols.len());
    println!("  isolated files  {}", files.len());
    println!("  hollow stubs    {}", hollow.len());
    println!();
    for item in symbols.iter().take(25) {
        println!(
            "  {}::{}  {}",
            item["relpath"].as_str().unwrap_or(""),
            item["name"].as_str().unwrap_or(""),
            item["signature"].as_str().unwrap_or("")
        );
    }
    if !files.is_empty() {
        println!();
        println!("Isolated files");
        for item in files.iter().take(20) {
            println!(
                "  {}  ({} loc)",
                item["relpath"].as_str().unwrap_or(""),
                item["loc"]
            );
        }
    }
}

pub fn print_slice(data: &Value) {
    println!("Behavior slice");
    println!();
    kv("query", &data["query"]);
    if let Some(err) = data["error"].as_str() {
        println!("  {:<18} {}", "error", err);
        if let Some(candidates) = data["candidates"].as_array() {
            if !candidates.is_empty() {
                println!();
                println!("Candidates (use path::name)");
                for item in candidates.iter().take(20) {
                    println!(
                        "  {}::{}",
                        item["relpath"].as_str().unwrap_or(""),
                        item["name"].as_str().unwrap_or("")
                    );
                }
            }
        }
        return;
    }
    kv("blast radius", &data["blast_radius"]);
    if let Some(effects) = data["effects"].as_array() {
        let tags: Vec<&str> = effects.iter().filter_map(|e| e.as_str()).collect();
        if !tags.is_empty() {
            println!("  {:<18} {}", "effects", tags.join(", "));
        }
    }
    if let Some(unresolved) = data["unresolved_callees"].as_array() {
        let tags: Vec<&str> = unresolved.iter().filter_map(|e| e.as_str()).take(12).collect();
        if !tags.is_empty() {
            println!("  {:<18} {}", "unresolved", tags.join(", "));
        }
    }
    println!();
    if let Some(symbols) = data["symbols"].as_array() {
        println!("Symbols");
        for item in symbols {
            if let Some(s) = item.as_str() {
                println!("  {s}");
            } else {
                println!(
                    "  {}::{}",
                    item["relpath"].as_str().unwrap_or(""),
                    item["name"].as_str().unwrap_or("")
                );
            }
        }
    }
    if let Some(files) = data["files"].as_array() {
        println!();
        println!("Files");
        for item in files {
            if let Some(s) = item.as_str() {
                println!("  {s}");
            }
        }
    }
}

pub fn print_cluster(data: &Value) {
    let clones = data["clones"].as_array().cloned().unwrap_or_default();
    println!("Clusters");
    println!("  clone pairs  {}", clones.len());
    println!();
    for item in clones.iter().take(20) {
        println!(
            "  {:.2}  {}  ↔  {}",
            item["score"].as_f64().unwrap_or(0.0),
            item["a"].as_str().or_else(|| item["left"].as_str()).unwrap_or(""),
            item["b"].as_str().or_else(|| item["right"].as_str()).unwrap_or("")
        );
    }
    if let Some(near) = data["similar"].as_array() {
        if !near.is_empty() {
            println!();
            println!("Similar");
            for item in near {
                println!(
                    "  {:.3}  {}::{}",
                    item["score"].as_f64().unwrap_or(0.0),
                    item["relpath"].as_str().unwrap_or(""),
                    item["name"].as_str().unwrap_or("")
                );
            }
        }
    }
}

pub fn print_history(data: &Value) {
    println!("Git history");
    println!();
    let summary = data.get("summary").unwrap_or(data);
    kv("enabled", &summary["enabled"]);
    if summary.get("commits").is_some() {
        kv("commits", &summary["commits"]);
    }
    if let Some(reason) = summary["reason"].as_str() {
        println!("  {:<18} {reason}", "reason");
    }
}

pub fn print_list(items: &[Value]) {
    if items.is_empty() {
        println!("(no rows)");
        return;
    }
    if items[0].get("why").is_some() && items[0].get("name").is_some() {
        println!("Reading order");
        for (i, item) in items.iter().enumerate() {
            println!(
                "  {:>2}. {}::{}  —  {}",
                i + 1,
                item["relpath"].as_str().unwrap_or(""),
                item["name"].as_str().unwrap_or(""),
                item["why"].as_str().unwrap_or("")
            );
        }
        return;
    }
    if items[0].get("kind").is_some() && items[0].get("detail").is_some() {
        println!("Flags ({})", items.len());
        for item in items {
            let loc = match (item["relpath"].as_str(), item["name"].as_str()) {
                (Some(p), Some(n)) if !n.is_empty() => format!("{p}::{n}"),
                (Some(p), _) => p.to_string(),
                _ => String::new(),
            };
            println!(
                "  {:<22} {}  {}",
                item["kind"].as_str().unwrap_or(""),
                loc,
                item["detail"].as_str().unwrap_or("")
            );
        }
        return;
    }
    for item in items {
        if let (Some(p), Some(n)) = (item["relpath"].as_str(), item["name"].as_str()) {
            println!("  {p}::{n}");
        } else {
            println!("  {item}");
        }
    }
}

fn kv(label: &str, value: &Value) {
    if value.is_null() {
        return;
    }
    let text = match value {
        Value::String(s) => s.clone(),
        Value::Number(n) => n.to_string(),
        Value::Bool(b) => b.to_string(),
        other => other.to_string(),
    };
    println!("  {label:<18} {text}");
}

File: src/dued/display.py
from __future__ import annotations

import json
from pathlib import Path

from rich.console import Console
from rich.table import Table

from dued import __version__
from dued.paths import db_path, report_root

console = Console()


def banner(repo: Path, command: str, with_embed: bool, with_git: bool, model: str) -> None:
    console.print(f"[bold]dued[/bold] {__version__}")
    console.print(f"repo     {repo}")
    console.print(f"index    {db_path(repo)}")
    console.print(f"reports  {report_root(repo)}")
    console.print(f"command  {command}")
    steps = ["walk source files", "parse + metrics", "build call graph"]
    if with_git:
        steps.append("git history")
    steps.append("rank + name health")
    if with_embed:
        if model == "stub":
            steps.append("embed symbols (stub vectors)")
        else:
            steps.append("embed symbols (Jina, local ONNX)")
    else:
        steps.append("skip embeddings")
    steps.append("write reports")
    console.print("plan     " + " → ".join(steps))
    if with_embed and model != "stub":
        console.print("note     first Jina load downloads ONNX weights (can take several minutes)")
        console.print("note     later runs reuse the Hugging Face cache")
    console.print("note     use --json for machine output; [bold]dued report[/bold] to re-read results")
    console.print()


def emit(data: object, as_json: bool) -> None:
    if as_json:
        # Plain stdout: Rich soft-wrap inserts raw newlines inside JSON strings.
        print(json.dumps(data, indent=2, default=str), flush=True)
        return
    if isinstance(data, dict):
        if "report" in data and "files" in data:
            print_analyze(data)
            return
        if "reading_order" in data and "issues" in data:
            print_brief(data, None)
            return
        if "symbols" in data and "hollow" in data and "edges" not in data:
            print_dead(data)
            return
        if "blast_radius" in data or ("effects" in data and "symbols" in data and "files" in data):
            print_slice(data)
            return
        if "clones" in data and "clusters" in data:
            print_cluster(data)
            return
    if isinstance(data, list):
        print_list(data)
        return
    console.print(data)


def print_analyze(data: dict) -> None:
    console.print("[bold]Scan complete[/bold]\n")
    for key, label in (
        ("files", "files"),
        ("symbols", "symbols"),
        ("edges", "edges"),
        ("parsed", "parsed this run"),
        ("reused", "reused files"),
        ("issues", "issues"),
        ("hollow", "hollow stubs"),
        ("clones", "clones"),
        ("mismatches", "mismatches"),
        ("model", "model"),
        ("elapsed_seconds", "elapsed seconds"),
    ):
        if key in data:
            console.print(f"  {label:<18} {data[key]}")
    inv = data.get("inventory") or {}
    langs = inv.get("languages") or []
    if langs:
        text = ", ".join(f"{row.get('language')} ({row.get('n')} files)" for row in langs)
        console.print(f"  {'languages':<18} {text}")
    git = data.get("git") or {}
    if git.get("enabled"):
        console.print(f"  {'git':<18} {git.get('commits')} commits")
    report = data.get("report")
    if report:
        dest = Path(str(report))
        console.print()
        console.print("[bold]Reports[/bold]")
        console.print(f"  HTML     {dest / 'report.html'}")
        console.print(f"  folder   {dest}")
        console.print()
        console.print("[bold]Explore[/bold]")
        console.print("  open the HTML file in a browser")
        console.print("  dued report              re-print this summary from disk")
        console.print("  dued rank                reading order")
        console.print("  dued issues              flagged problems")
        console.print("  dued dead                unused symbols and files")
        console.print("  dued names               name-health flags")
        console.print("  dued cluster             clones")
        console.print("  dued slice <symbol>      behavior slice")


def print_brief(data: dict, html: Path | None) -> None:
    console.print("[bold]dued report[/bold]\n")
    console.print(f"  {'repo':<18} {data.get('repo', '')}")
    console.print(f"  {'files':<18} {data.get('files', '')}")
    console.print(f"  {'symbols':<18} {data.get('symbols', '')}")
    langs = data.get("languages") or []
    if langs:
        text = ", ".join(f"{row.get('language')}={row.get('n')}" for row in langs)
        console.print(f"  {'languages':<18} {text}")
    console.print()
    table = Table(title="Reading order")
    table.add_column("#", style="dim")
    table.add_column("symbol")
    table.add_column("why")
    for i, item in enumerate(data.get("reading_order") or [], start=1):
        table.add_row(str(i), f"{item.get('relpath')}::{item.get('name')}", str(item.get("why") or ""))
    console.print(table)
    issues = data.get("issues") or []
    if issues:
        console.print()
        itable = Table(title=f"Issues ({len(issues)})")
        itable.add_column("kind")
        itable.add_column("where")
        itable.add_column("detail")
        for item in issues[:15]:
            name = item.get("name") or ""
            loc = f"{item.get('relpath') or ''}::{name}" if name else str(item.get("relpath") or "")
            itable.add_row(str(item.get("kind") or ""), loc, str(item.get("detail") or ""))
        console.print(itable)
    if html is not None:
        console.print()
        console.print("[bold]HTML report[/bold]")
        console.print(f"  {html}")
        console.print("  open that file in a browser to search and sort the full index")
        console.print("  JSON tables are in the data/ folder next to the HTML file")


def print_dead(data: dict) -> None:
    symbols = data.get("symbols") or []
    files = data.get("files") or []
    hollow = data.get("hollow") or []
    console.print("[bold]Dead code[/bold]")
    console.print(f"  unused symbols  {len(symbols)}")
    console.print(f"  isolated files  {len(files)}")
    console.print(f"  hollow stubs    {len(hollow)}")
    console.print()
    for item in symbols[:25]:
        console.print(f"  {item.get('relpath')}::{item.get('name')}  {item.get('signature') or ''}")


def print_slice(data: dict) -> None:
    console.print("[bold]Behavior slice[/bold]\n")
    console.print(f"  {'query':<18} {data.get('query', '')}")
    if data.get("error"):
        console.print(f"  {'error':<18} {data.get('error')}")
        candidates = data.get("candidates") or []
        if candidates:
            console.print()
            console.print("Candidates (use path::name)")
            for item in candidates[:20]:
                console.print(f"  {item.get('relpath')}::{item.get('name')}")
        return
    console.print(f"  {'blast radius':<18} {data.get('blast_radius', '')}")
    effects = data.get("effects") or []
    if effects:
        tags = ", ".join(str(x) for x in effects)
        console.print(f"  {'effects':<18} {tags}")
    unresolved = data.get("unresolved_callees") or []
    if unresolved:
        tags = ", ".join(str(x) for x in unresolved[:12])
        console.print(f"  {'unresolved':<18} {tags}")


def print_cluster(data: dict) -> None:
    clones = data.get("clones") or []
    console.print("[bold]Clusters[/bold]")
    console.print(f"  clone pairs  {len(clones)}")


def print_list(items: list) -> None:
    if not items:
        console.print("(no rows)")
        return
    first = items[0] if isinstance(items[0], dict) else {}
    if first.get("why") and first.get("name"):
        table = Table(title="Reading order")
        table.add_column("#", style="dim")
        table.add_column("symbol")
        table.add_column("why")
        for i, item in enumerate(items, start=1):
            table.add_row(str(i), f"{item.get('relpath')}::{item.get('name')}", str(item.get("why") or ""))
        console.print(table)
        return
    if first.get("kind") and first.get("detail"):
        table = Table(title=f"Flags ({len(items)})")
        table.add_column("kind")
        table.add_column("where")
        table.add_column("detail")
        for item in items:
            name = item.get("name") or ""
            loc = f"{item.get('relpath') or ''}::{name}" if name else str(item.get("relpath") or "")
            table.add_row(str(item.get("kind") or ""), loc, str(item.get("detail") or ""))
        console.print(table)
        return
    for item in items:
        if isinstance(item, dict) and item.get("relpath") and item.get("name"):
            console.print(f"  {item['relpath']}::{item['name']}")
        else:
            console.print(f"  {item}")
